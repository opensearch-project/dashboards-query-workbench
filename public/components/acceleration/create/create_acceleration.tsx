/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  EuiButton,
  EuiButtonEmpty,
  EuiFlexGroup,
  EuiFlexItem,
  EuiFlyout,
  EuiFlyoutBody,
  EuiFlyoutFooter,
  EuiFlyoutHeader,
  EuiSpacer,
} from '@elastic/eui';
import React, { useState } from 'react';
import { ACCELERATION_TIME_INTERVAL } from '../../../../common/constants';
import { CreateAccelerationForm } from '../../../../common/types/';
import { DefineIndexOptions } from '../selectors/define_index_options';
import { IndexSettingOptions } from '../selectors/index_setting_options';
import { AccelerationDataSourceSelector } from '../selectors/source_selector';
import { accelerationQueryBuilder } from '../visual_editors/query_builder';
import { QueryVisualEditor } from '../visual_editors/query_visual_editor';
import { CautionBannerCallout } from './caution_banner_callout';
import { CreateAccelerationHeader } from './create_acceleration_header';

export interface CreateAccelerationProps {
  dataSource: string;
  setIsFlyoutVisible(visible: boolean): void;
  updateQueries: (query: string) => void;
}

export const CreateAcceleration = ({
  dataSource,
  setIsFlyoutVisible,
  updateQueries,
}: CreateAccelerationProps) => {
  const [accelerationFormData, setAccelerationFormData] = useState<CreateAccelerationForm>({
    dataSource: '',
    dataTable: '',
    database: '',
    dataTableFields: [],
    accelerationIndexType: 'skipping',
    skippingIndexQueryData: [],
    coveringIndexQueryData: [],
    materializedViewQueryData: {
      columnsValues: [],
      groupByTumbleValue: {
        timeField: '',
        tumbleWindow: 0,
        tumbleInterval: '',
      },
    },
    accelerationIndexName: '',
    primaryShardsCount: 5,
    replicaShardsCount: 1,
    refreshType: 'auto',
    checkpointLocation: undefined,
    refreshIntervalOptions: {
      refreshWindow: 1,
      refreshInterval: ACCELERATION_TIME_INTERVAL[1].value,
    },
  });

  const copyToEditor = () => {
    updateQueries(accelerationQueryBuilder(accelerationFormData));
  };

  return (
    <>
      <EuiFlyout
        ownFocus
        onClose={() => setIsFlyoutVisible(false)}
        aria-labelledby="flyoutTitle"
        size="m"
      >
        <EuiFlyoutHeader hasBorder>
          <CreateAccelerationHeader />
        </EuiFlyoutHeader>
        <EuiFlyoutBody>
          <CautionBannerCallout />
          <EuiSpacer size="l" />
          <AccelerationDataSourceSelector
            accelerationFormData={accelerationFormData}
            setAccelerationFormData={setAccelerationFormData}
          />
          <EuiSpacer size="xxl" />
          <IndexSettingOptions
            accelerationFormData={accelerationFormData}
            setAccelerationFormData={setAccelerationFormData}
          />
          <EuiSpacer size="xxl" />
          <DefineIndexOptions
            accelerationFormData={accelerationFormData}
            setAccelerationFormData={setAccelerationFormData}
          />
          <EuiSpacer size="m" />
          <QueryVisualEditor
            accelerationFormData={accelerationFormData}
            setAccelerationFormData={setAccelerationFormData}
          />
        </EuiFlyoutBody>
        <EuiFlyoutFooter>
          <EuiFlexGroup justifyContent="spaceBetween">
            <EuiFlexItem grow={false}>
              <EuiButtonEmpty
                iconType="cross"
                onClick={() => setIsFlyoutVisible(false)}
                flush="left"
              >
                Close
              </EuiButtonEmpty>
            </EuiFlexItem>
            <EuiFlexItem grow={false}>
              <EuiButton
                onClick={() => {
                  copyToEditor();
                  setIsFlyoutVisible(false);
                }}
                fill
              >
                Copy Query to Editor
              </EuiButton>
            </EuiFlexItem>
          </EuiFlexGroup>
        </EuiFlyoutFooter>
      </EuiFlyout>
    </>
  );
};
