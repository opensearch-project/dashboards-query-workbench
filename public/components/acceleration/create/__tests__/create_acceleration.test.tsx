/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { waitFor } from '@testing-library/dom';
import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import { CreateAcceleration } from '../create_acceleration';

describe('Create acceleration flyout components', () => {
  configure({ adapter: new Adapter() });

  it('renders acceleration flyout component with default options', async () => {
    const dataSource = '';
    const resetFlyout = jest.fn();
    const updateQueries = jest.fn();
    const wrapper = mount(
      (
        <CreateAcceleration
          dataSource={dataSource}
          resetFlyout={resetFlyout}
          updateQueries={updateQueries}
        />
      ) as React.ReactElement<any>
    );
    wrapper.update();
    await waitFor(() => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
