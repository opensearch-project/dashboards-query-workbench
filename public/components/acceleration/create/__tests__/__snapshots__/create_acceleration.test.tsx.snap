// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create acceleration flyout components renders acceleration flyout component with default options 1`] = `ReactWrapper {}`;
